### PR TITLE
Support for user-type actors

### PR DESCRIPTION
**Pull request by OleksiiOleksenko**
The PR adds basic support for userspace code and for transitions between kernel and user actors. Namely:
- mapping of sandbox pages to user space
- macros for switching from host to user, and back (both executor and model)
- setting of MSR_LSTAR
- replace R11 with R13 for storing htraces; R11 is now used for syscall/sysret macros

~~Known issues (to be fixed before merging):~~ All fixed now; ready to merge

- [x] model does not include changes to RCX and RFLAGS caused by macros
- [x] the target of `switch_u2h` is not configurable; it always lands into the fault handler (located at `TEMPLATE_DEFAULT_EXCEPTION_LANDING` within the main section template)
- [x] `measurement_start` and `measurement_end` macros cause a fault when executed by user-type actor

Minimal example test case:

```assembly
.intel_syntax noprefix
.test_case_enter:

# ----------------------  Actor 1 ----------------------
.section .data.main

.function_start:
    .macro.select_switch_u2h_target.main.function_fin:  # optional
    .macro.select_switch_h2u_target.actor2.function_1:  # non-optional
    .macro.switch_h2u.actor2:

.function_fin:
    nop
# ------------------------------------------------------

# ----------------------  Actor 2 ----------------------
.section .data.actor2
.function_1:
    # memory access to check that we can observe user loads
    mov rax, qword ptr [r14 + 0x200]
    .macro.switch_u2h.main:
# ------------------------------------------------------

.test_case_exit:
```

Config file:
```yaml
instruction_set: x86-64

inputs_per_class: 2

executor_repetitions: 10
executor_warmups: 1
executor_mode: P+P

# uncomment to test correctness of the model
# fuzzer: architectural

program_size: 32
avg_mem_accesses: 16
min_bb_per_function: 1
max_bb_per_function: 1

enable_speculation_filter: false
enable_observation_filter: false
enable_fast_path_model: false
enable_fast_path_executor: false
ignore_flaky_violations: true

actor:
  - name: "main"
  - mode: "host"

actor:
  - name: "actor2"
  - mode: "user"

contract_observation_clause: ct
contract_execution_clause:
  - seq

color: true
logging_modes:
  - info
  - stat
  - dbg_traces  # uncomment to see htraces
  - dbg_traces_no_ctr   # remove to see ctraces
  # - dbg_model  # uncomment to debug execution on the model
```

FYI @flaviens @bkoepf @janahofmann

### CODE REVIEW

File: src/generator.py [hunk 1]
```diff
@@ -185,6 +185,8 @@ def create_actors(self, test_case: TestCase):
                 mode = ActorMode.HOST
             elif desc['mode'] == "guest":
                 mode = ActorMode.GUEST
+            elif desc['mode'] == "user":
```
Comment (flaviens) [inline src/generator.py:188]: As discussed, we should keep some orthogonality between privilege (user vs supervisor) and world (host vs guest)

File: src/interfaces.py [hunk 2]
```diff
@@ -143,6 +143,7 @@ def get_sandbox_addr(sandbox_base: int, name: str):
 class ActorMode(Enum):
     HOST = 0
```
Comment (flaviens) [inline src/interfaces.py:144]: Same as above regarding orthogonality.

File: src/util.py [hunk 3]
```diff
@@ -134,6 +134,7 @@ class Logger:
     dbg_model: bool = False
     dbg_coverage: bool = False
     dbg_generator: bool = False
+    dbg_traces_no_ctr: bool = False
```
Comment (flaviens) [inline src/util.py:137]: Probably having only "positive" flags would make things a bit clearer

File: src/x86/executor/include/actor.h [hunk 4]
```diff
@@ -14,6 +14,7 @@ typedef uint64_t actor_mode_t;
 enum {
     MODE_HOST = 0,
```
Comment (flaviens) [inline src/x86/executor/include/actor.h:15]: Same thing about privileges vs world

File: src/x86/executor/sandbox_manager.c [hunk 5]
```diff
@@ -86,6 +86,22 @@ int allocate_sandbox(void)
     return 0;
 }
 
+/// @brief Returns the number of pages allocated for the sandbox, including util area, code and data
+/// @param void
+/// @return number of pages; -1 on error
+int get_n_sandbox_pages(void) {
```
Comment (flaviens) [inline src/x86/executor/sandbox_manager.c:92]: I think the name is a bit misleading (to me it sounds like "allocate n pages for the sandbox"). How about `get_sandbox_size_pages` or something alike?

File: src/x86/executor/test_case_parser.c [hunk 6]
```diff
@@ -161,6 +161,19 @@ static int __batch_tc_parsing_end(void)
         PRINT_ERRS("__batch_tc_parsing_end", "Symbol table does not contain main function\n");
         return -1;
     }
+
```
Comment (flaviens) [inline src/x86/executor/test_case_parser.c:164]: Same thing, potentially, as privil vs world

**Review by flaviens — COMMENTED** [review]
Cool stuff!

File: src/generator.py (hunk 1 above)
Comment (OleksiiOleksenko) [reply src/generator.py:188]: Agree. Should we update the keyword to "host-user" and "host-kernel" then?

File: src/util.py (hunk 3 above)
Comment (OleksiiOleksenko) [reply src/util.py:137]: good point. on it

File: src/generator.py (hunk 1 above)
Comment (flaviens) [reply src/generator.py:188]: Having 2 distinct keys in the dict would feel more intuitive to me but as you prefer :+1:

File: src/util.py (hunk 3 above)
Comment (OleksiiOleksenko) [reply src/util.py]: fixed

File: src/x86/executor/sandbox_manager.c (hunk 5 above)
Comment (OleksiiOleksenko) [reply src/x86/executor/sandbox_manager.c]: fixed

File: src/generator.py (hunk 1 above)
Comment (OleksiiOleksenko) [reply src/generator.py]: done

**Pull request merged**